libportal/notification.c

```diff
@@ -119,6 +191,283 @@ call_done_cb (GObject      *source,
+
+  stream_in = G_INPUT_STREAM (g_file_read_finish (G_FILE (source_object), res, &error));
+
+  stream_out = g_memory_output_stream_new_resizable ();
+
+  g_output_stream_splice_async (g_steal_pointer (&stream_out),
+                                stream_in,
+                                G_OUTPUT_STREAM_SPLICE_CLOSE_SOURCE | G_OUTPUT_STREAM_SPLICE_CLOSE_TARGET,
+                                g_task_get_priority (task),
+                                g_task_get_cancellable (task),
+                                splice_for_bytes_icon_cb,
+                                g_object_ref (task));
+}
+
+static gboolean
+parse_icon_property (GTask            *task,
+                     GVariant         *serialized_icon)
+{
+  XdpPortal *portal = g_task_get_source_object (task);
+  CallData *data = g_task_get_task_data (task);
+  g_autoptr(GIcon) icon = NULL;
+
+  icon = g_icon_deserialize (serialized_icon);
+
+  if (G_IS_BYTES_ICON (icon) && portal->notification_interface_version > 1)
+    {
+      g_autoptr(GError) error = NULL;
+      g_autofd int fd = -1;
+      int fd_in;
+
+      fd = bytes_to_memfd ("notification-media",
+                           g_bytes_icon_get_bytes (G_BYTES_ICON (icon)),
+                           &error);
+      if (fd == -1)
+        {
+          g_task_return_error (task, g_steal_pointer (&error));
+          return TRUE;
+        }
+
+      fd_in = g_unix_fd_list_append (data->fd_list, fd, &error);
+      if (fd_in == -1)
+        {
+          g_task_return_error (task, g_steal_pointer (&error));
+          return TRUE;
+        }
+
+      g_variant_builder_add (data->builder, "{sv}", "icon",
+                             g_variant_new ("(sv)", "file-descriptor", g_variant_new_handle (fd_in)));
+
+      return FALSE;
+    }
+  else if (G_IS_FILE_ICON (icon) && portal->notification_interface_version > 1)
+    {
+      g_variant_builder_open (data->builder, G_VARIANT_TYPE ("{sv}"));
+      g_variant_builder_add (data->builder, "s", "icon");
+      g_variant_builder_open (data->builder, G_VARIANT_TYPE_VARIANT);
+
+      g_file_read_async (g_file_icon_get_file (G_FILE_ICON (icon)),
+                         g_task_get_priority (task),
+                         g_task_get_cancellable (task),
+                         file_read_cb,
+                         g_object_ref (task));
+
+      return TRUE;
+    }
+  else if (G_IS_FILE_ICON (icon))
+    {
+      g_file_read_async (g_file_icon_get_file (G_FILE_ICON (icon)),
+                         g_task_get_priority (task),
+                         g_task_get_cancellable (task),
+                         file_read_for_bytes_icon_cb,
+                         g_object_ref (task));
+
+      return TRUE;
+    }
+  else
+    {
+      g_variant_builder_add (data->builder, "{sv}", "icon", serialized_icon);
+
+      return FALSE;
+    }
+}
+
+static gboolean
+parse_sound_property (GTask            *task,
+                      GVariant         *serialized_sound)
```

Comment:
It's a bit unnecessary to pass the task all the way here when all you need it the builder and return an error.

Reply:
`parse_sound_property()` does async and It needs to be passed task to `file_read_cb()`

Reply:
I think it would be more understandable if you first collect all files that you want to read, then do the reading, then do the "parsing".

Reply:
yes, that's what i had initially. But for v1 we need to convert `GFileIcon` to `GBytesIcon` which has the icon content inline. The I think the only two option for that is to parse the vardicts twice, the second time adding the bytes, or modify the `GVariant` inplace, which sounds a bad thing to do.

Reply:
Iterating over the GVariant once to figure out which files need to be read and then once more to build the new variant that is handed to dbus sounds completely fine to me.

Reply:
I kept the parsing function still async, but without restarting the loop. The parse walks the vardict once and completes once the async loading of the icon/sound has finished.